### ISSUE TITLE
LayoutAnimations `entering` and `exiting` not working

### ISSUE THREAD

**Issue by mrousavy** [software-mansion/react-native-reanimated#5395]
### Description

Hey yall,

I tried playing around with LayoutAnimations and I love the idea - they're so simple to implement, amazing work!

https://github.com/software-mansion/react-native-reanimated/assets/15199031/3fac8af4-25c7-4482-9a4d-877fe70c30c0

I did notice that there are a few bugs though.

I created a repro here that showcases those bugs: https://github.com/software-mansion/react-native-reanimated/pull/5394

## 1. Exiting not running

The `exiting` animation is not running in the given example.

## 2. Custom `useAnimatedStyle` not running on first mount

On the first mount, the `useAnimatedStyle` for the second card is not properly applied. I think this is a race condition where the LayoutAnimation takes precedence over the `useAnimatedStyle`.

## 3. Some items not showing

Sometimes the items just don't show up.

## 4. `layout` instead of `useAnimatedStyle` does not work

When using `layout` instead of manually using `useAnimatedStyle` it will not consider properties in the `transform` style.

This is how I animate the slide & rotate to the right now:

```tsx
const rotation = useDerivedValue(() => withSpring(index * 15), [index]);

const animatedStyle = useAnimatedStyle(() => {
  return {
    zIndex: -index,
    backgroundColor: item.color,
    transform: [
      {
        translateX: interpolate(rotation.value, [0, 100], [0, 100]),
      },
      {
        translateY: interpolate(rotation.value, [0, 100], [0, -30]),
      },
      {
        scale: interpolate(rotation.value, [0, 100], [1, 0.7]),
      },
      {
        rotate: `${rotation.value}deg`,
      },
    ],
  };
}, [rotation, index, item]);

return (
  <Animated.View
    entering={Entering}
    exiting={Exiting}
    style={[styles.item, animatedStyle]}
  />
);
```

This is how I want to do it:

```tsx
const style = useMemo(() => {
  return {
    zIndex: -index,
    backgroundColor: item.color,
    transform: [
      {
        translateX: x,
      },
      {
        translateY: y,
      },
      {
        scale: scale,
      },
      {
        rotate: `${rotation}deg`,
      },
    ],
  };
}, [rotation, index, item]);

return (
  <Animated.View
    entering={Entering}
    exiting={Exiting}
    layout={Move} // <-- just a default Layout.springify() animation that should animate the props in transform: [...]
    style={[styles.item, style]}
  />
);
```

Let me know if you need more details!

### Steps to reproduce

https://github.com/software-mansion/react-native-reanimated/pull/5394 <-- open the new page

### Snack or a link to a repository

https://github.com/software-mansion/react-native-reanimated/pull/5394

### Reanimated version

3.5.0 (main)

### React Native version

0.72.6

### Platforms

Android, iOS

### JavaScript runtime

Hermes

### Workflow

React Native

### Architecture

Paper (Old Architecture)

### Build type

Debug mode

### Device

iOS simulator

### Device model

iPhone 15 Pro Simulator

### Acknowledgements

Yes

**Comment by Iannery** [comment]
Same error here.

**Comment by tomekzaw** [comment]
Hey @mrousavy, thanks again for reporting this issue.

> ### 1. Exiting not running

The exiting animation is running (you can substitute `ZoomOut` with `FadeOut` and it will work correctly), but does not account for transforms. In this case, it's overlapped by other cards while exiting. If you change the opacity of the cards, you can see exactly what happens:

https://github.com/software-mansion/react-native-reanimated/assets/20516055/ef1dbeca-2b0d-49fc-9760-28d3f1c58608

> ### 2. Custom `useAnimatedStyle` not running on first mount

Not really sure what this means, could you provide more details on what's the actual and expected behavior? One problem I can clearly see is that during the first render the position of the cards does not account for header height; this is a regression caused by some changes in react-native-screens, here's how it looks like:

<table>
<thead>
<tr>
<th>Without screen</th>
<th>Inside screen</th>
</tr>
</thead>
<tbody>
<tr>
<td><video src="https://github.com/software-mansion/react-native-reanimated/assets/20516055/5e012e9f-b468-45ec-93e4-e769439bfb72" /></td>
<td><video src="https://github.com/software-mansion/react-native-reanimated/assets/20516055/a7519e25-6936-492d-8e06-cd881dacf991" /></td>
</tr>
</tbody>
</table>

> ### 3. Some items not showing

I was able to reproduce this problem but it looks like it's caused by incorrect implementation of `getRandomColor` in your example. Sometimes it returns an invalid hex color (e.g. `#1d7ea`). Here's a fix that solved it for me:

```diff
 function getRandomColor(): string {
-  return '#' + Math.floor(Math.random() * 16777215).toString(16);
+  return '#' + Math.floor(Math.random() * 16777215).toString(16).padStart(6, '0');
 }
```

> ### 4. `layout` instead of `useAnimatedStyle` does not work

That's true, currently `entering`, `layout` and `exiting` animations simply overwrite existing transforms. It would be really cool if `layout` supported animating transforms as well. The current behavior is the following: if a component changes its Yoga layout (position OR size), we run the animation. Technically, transforms don't affect the layout of the component so in this case the layout does not change at all.

**Comment by mrousavy** [comment]
Thanks for the detailed insights Tomasz, really interesting!

Regarding "2. Custom useAnimatedStyle not running on first mount":

I am not sure what the issue there was as I am not able to reproduce it anymore - but in my app the back card was not rotated to the right when the app first starts, it was just behind the first card - no rotations applied. So it was invisible to the user.

I'll try to reproduce that again when I get the chance!

**Comment by computerjazz** [comment]
Hey all, I've opened previous issues with repros for some of these:
- First mount animations: https://github.com/software-mansion/react-native-reanimated/issues/3296
- Missing exiting animations: https://github.com/software-mansion/react-native-reanimated/issues/5349

**Comment by mohamine18** [comment]
**Problem:**
I encountered a similar issue where combining two animated styles using the `style={[animatedStyle1, animatedStyle2]}` syntax worked as expected on Android but not on iOS. One style was for sliding a component, and the other was for animating the backgroundColor of the same component.

**Solution Attempt:**
To address the issue on iOS, I found a workaround by wrapping each animated style in a separate Animated.View component. This involved using two separate views with their respective styles: `<Animated.View style={animatedStyle1}><Animated.View style={animatedStyle2}>...</Animated.View></Animated.View>.` While this resolved the problem on iOS, it feels like a less-than-ideal solution.

**Suggested Improvement:**
It would be great if the combined style syntax `style={[animatedStyle1, animatedStyle2]}` could work consistently across both Android and iOS platforms. This would enhance the developer experience by providing a more straightforward and cross-platform solution.

Additional Information:

React Native Version: 0.72.6
React Native Reanimated: 3.3.0
Expo SDK: 49
Platform: iOS 17.2
